bpo-39429: Add a new "Python Development Mode" doc page (#18132)
python/cpython: Doc/library/devmode.rst

.. _devmode:

Python Development Mode
=======================

.. versionadded:: 3.7

> The Python Development Mode introduces additional runtime checks which are too

Reviewer: ```suggestion
The Python Development Mode introduces additional runtime checks that are too
```

Reviewer 2: I think `which` is correct here, but it's probably debatable.  https://www.grammarly.com/blog/which-vs-that/

Reviewer: Upon further consideration, I think either of the two are probably reasonable here. "that" just sounded a bit better in my mind when I was reading over it.

expensive to be enabled by default. It should not be more verbose than the

> default if the code is correct: new warnings are only emitted when an issue is

Reviewer 2: The colon should probably be a semi-colon.

detected.

> It can be enabled using :option:`-X dev <-X>` command line option or by setting

Reviewer: ```suggestion
It can be enabled using the :option:`-X dev <-X>` command line option or by setting
```

> :envvar:`PYTHONDEVMODE` environment variable to ``1``.

Reviewer: ```suggestion
the :envvar:`PYTHONDEVMODE` environment variable to ``1``.
```

Effects of the developer mode
=============================

Enabling the development mode has multiple effects:

* Add ``default`` :ref:`warning filter <describing-warning-filters>`, as
  :option:`-W default <-W>` command line option. The following warnings are
  shown, whereas they are ignored by the default warning filters:

  * :exc:`DeprecationWarning`;
  * :exc:`ImportWarning`;
  * :exc:`PendingDeprecationWarning`;

>   * :exc:`ResourceWarning`.

Reviewer: Hmm, the wording of this section isn't entirely clear to me. Is it explaining that the listed warnings are *not* filtered in development mode, so they're shown instead of hidden? Also, that to re-use the default warning filters, a new command-line option "-W default" has been added?

I had to read over this part a few times, and I'm still not 100% certain that I'm quite getting it.

If I'm correctly understanding it, I'd recommend restructuring it a bit to make that more clear (might need to adjust line width):
```
Normally, the following warnings are filtered by the default 
:ref:`warning filters <describing-warning-filters>`:

  * :exc:`DeprecationWarning`;
  * :exc:`ImportWarning`;
  * :exc:`PendingDeprecationWarning`;
  * :exc:`ResourceWarning`.

In development mode, these warnings are no longer filtered. To re-enable the default
warning filters, use the new :option:`-W default <-W>` command line option.
```
(Correct me if I'm misunderstanding the section)

Author: I rewrote this section.

>   Use :option:`-W error <-W>` command line option or set

Reviewer: ```suggestion
  Use the :option:`-W error <-W>` command line option or set
```

>   :envvar:`PYTHONWARNINGS` environment variable to ``error`` to treat warnings

Reviewer: ```suggestion
  the :envvar:`PYTHONWARNINGS` environment variable to ``error`` to treat warnings
```

  as errors.

* Install debug hooks on memory allocators to check for:

  * buffer underflow;
  * buffer overflow
  * memory allocator API violation;

>   * usage of the GIL.

Reviewer 2: I'm not sure any of these should end in a semi-colon, but if you want to keep that style, you're missing one.  :-)

Reviewer: I've seen this style used before in a few different areas throughout the documentation, so I think it's okay. But, it should definitely be consistent within the same listing of bullet points. (:

  See the :c:func:`PyMem_SetupDebugHooks` C function. Similar to setting

>   :envvar:`PYTHONMALLOC` environment variable to ``debug``.

Reviewer: ```suggestion
  the :envvar:`PYTHONMALLOC` environment variable to ``debug``.
```

>   To enable the development without installing debug hooks on memory

Reviewer: ```suggestion
  To enable the development mode without installing debug hooks on memory
```

>   allocators, set :envvar:`PYTHONMALLOC` environment variable to ``default``.

Reviewer: ```suggestion
  allocators, set the :envvar:`PYTHONMALLOC` environment variable to ``default``.
```

* Call :func:`faulthandler.enable` at Python startup to enable the

>   :mod:`faulthandler` module: install handlers for the :const:`SIGSEGV`,

Reviewer 2: Does calling `faulthandler.enable` install the handlers?  If so, it might be better to say `which installs` instead of `: install`.

Author: I phrased it differently.

  :const:`SIGFPE`, :const:`SIGABRT`, :const:`SIGBUS` and :const:`SIGILL`

>   signals to dump the Python traceback. Similar to :option:`-X faulthandler

Reviewer: ```suggestion
  signals to dump the Python traceback. Similar to the :option:`-X faulthandler
```

>   <-X>` command line option or setting :envvar:`PYTHONFAULTHANDLER` environment

Reviewer: ```suggestion
  <-X>` command line option or setting the :envvar:`PYTHONFAULTHANDLER` environment
```

  variable to ``1``.

* Enable :ref:`asyncio debug mode <asyncio-debug-mode>`. Similar to setting

>   :envvar:`PYTHONASYNCIODEBUG` environment variable to ``1``.

Reviewer: ```suggestion
  the :envvar:`PYTHONASYNCIODEBUG` environment variable to ``1``.
```

> * Check *encoding* and *errors* arguments on string encoding and decoding

Reviewer: ```suggestion
* Check the *encoding* and *errors* arguments for string encoding and decoding
```

  operations. Examples: :func:`open`, :meth:`str.encode` and
  :meth:`bytes.decode`.

> * :class:`io.IOBase` destructor logs ``close()`` exceptions.

Reviewer: ```suggestion
* The destructor for :class:`io.IOBase` logs ``close()`` exceptions.
```

* Set the :attr:`~sys.flags.dev_mode` attribute of :attr:`sys.flags` to
  ``True``.

The development mode does not enable the :mod:`tracemalloc` module by default,

> because the overhead (performance and memory) would be too important. Enabling

Reviewer: ```suggestion
because the overhead cost (to performance and memory) would be too detrimental. Enabling
```

the :mod:`tracemalloc` module provides additional information on the origin
of some errors. For example, :exc:`ResourceWarning` logs the traceback where
the resource was allocated, and a buffer overflow error logs the traceback
where the memory block was allocated.

> The development mode does not prevent the :option:`-O` command line option to

Reviewer: ```suggestion
The development mode does not prevent the :option:`-O` command line option from
```

> remove :keyword:`assert` statements nor to set :const:`__debug__` to ``False``.

Reviewer: ```suggestion
removing :keyword:`assert` statements nor from setting :const:`__debug__` to ``False``.
```

.. versionchanged:: 3.8

>    Log ``close()`` exceptions in :class:`io.IOBase` destructor.

Reviewer: ```suggestion
   Log ``close()`` exceptions in the :class:`io.IOBase` destructor.
```

ResourceWarning Example
=======================

> Example of a script counting the number of lines of the text file specified on

Reviewer: ```suggestion
Example of a script counting the number of lines of the text file specified in
```

the command line::

    import sys

    def main():
        fp = open(sys.argv[1])
        nlines = len(fp.readlines())
        print(nlines)

    if __name__ == "__main__":
        main()

The script does not close the file explicitly, but Python doesn't complain by

> default. Example with README.txt which is made 269 lines:

Reviewer: ```suggestion
default. Example using README.txt, which has 269 lines:
```

.. code-block:: shell-session

    $ python3 script.py README.txt
    269

Enabling the development mode displays a :exc:`ResourceWarning` warning:

.. code-block:: shell-session

    $ python3 -X dev script.py README.txt
    269
    script.py:9: ResourceWarning: unclosed file <_io.TextIOWrapper name='README.rst' mode='r' encoding='UTF-8'>
      main()
    ResourceWarning: Enable tracemalloc to get the object allocation traceback

> Enabling also :mod:`tracemalloc` shows the line where the file was opened:

Reviewer: ```suggestion
In addition, enabling :mod:`tracemalloc` shows the line where the file was opened:
```

.. code-block:: shell-session

    $ python3 -X dev -X tracemalloc=5 script.py README.rst
    269
    script.py:9: ResourceWarning: unclosed file <_io.TextIOWrapper name='README.rst' mode='r' encoding='UTF-8'>
      main()
    Object allocated at (most recent call last):
      File "script.py", lineno 9
        main()
      File "script.py", lineno 4
        fp = open(sys.argv[1])

The fix is to close explicitly the file. Example using a context manager::

    def main():
        with open(sys.argv[1]) as fp:
            nlines = len(fp.readlines())
        print(nlines)

Not closing a resource explicitly can leave a resource open for way longer than

> expected. It can cause severe issues at Python exit. It is bad in CPython, but

Reviewer: ```suggestion
expected; it can cause severe issues upon exit of the Python program. It is bad in CPython, but
```

Should we specify "interpreter" or "program" here? I think the former would be more accurate (at least from my understanding), particularly in the context of using subinterpreters, but the latter might be accurate enough for the purposes of this section.

Reviewer 2: Instead of `upon exit of the`, how about `upon exit from the`?

Reviewer: @csabella Hmm, or perhaps a third alternative: `upon exiting the`?

I think I read over this section a bit too many times, I'm starting to experience "[semantic sensation](https://en.wikipedia.org/wiki/Semantic_satiation)" (or I've just been awake for a bit too long). I'll leave it up to you.

        print(firstline.rstrip())
        os.close(fp.fileno())
        # fp object is destroyed implicitly

    main()

By default, Python does not emit any warning:

.. code-block:: shell-session

    $ python3 script.py
    import os

The development mode shows a :exc:`ResourceWarning` and logs a "Bad file
descriptor" error when finalizing the file object:

.. code-block:: shell-session

    $ python3 script.py
    import os
    script.py:10: ResourceWarning: unclosed file <_io.TextIOWrapper name='script.py' mode='r' encoding='UTF-8'>
      main()
    ResourceWarning: Enable tracemalloc to get the object allocation traceback
    Exception ignored in: <_io.TextIOWrapper name='script.py' mode='r' encoding='UTF-8'>
    Traceback (most recent call last):
      File "script.py", line 10, in <module>
        main()
    OSError: [Errno 9] Bad file descriptor

``os.close(fp.fileno())`` closes the file descriptor. When the file object
finalizer tries to close the file descriptor again, it fails with the ``Bad
file descriptor`` error. A file descriptor must be closed only once: closing it
twice can lead to a crash in the worst case (:issue:`18748` in a crash

> example).

Reviewer: I'd recommend the following rephrase:

```
A file descriptor must be closed only once. In the worst case scenario,
closing it twice can lead to a crash (see :issue:`18748` for an example of a crash).
```

> The fix is to remove ``os.close(fp.fileno())`` line, or open the file with

Reviewer: ```suggestion
The fix is to remove the ``os.close(fp.fileno())`` line, or open the file with
```

.. _devmode:

Python Development Mode
=======================

.. versionadded:: 3.7

The Python Development Mode introduces additional runtime checks that are too
expensive to be enabled by default. It should not be more verbose than the

> default if the code is correct: new warnings are only emitted when an issue is

Reviewer: ```suggestion
default if the code is correct; new warnings are only emitted when an issue is
```
(this applies @csabella's suggestion above)

detected.

It can be enabled using the :option:`-X dev <-X>` command line option or by
setting the :envvar:`PYTHONDEVMODE` environment variable to ``1``.

Effects of the Python Development Mode
======================================

Enabling the Python Development Mode is similar to the following command, but
with additional effects described below::

    PYTHONMALLOC=debug PYTHONASYNCIODEBUG=1 python3 -W default -X faulthandler

Effects of the Python Development Mode:

* Add ``default`` :ref:`warning filter <describing-warning-filters>`. The
  following warnings are shown:

  * :exc:`DeprecationWarning`
  * :exc:`ImportWarning`
  * :exc:`PendingDeprecationWarning`
  * :exc:`ResourceWarning`

>   Normally, the following warnings are filtered by the default :ref:`warning

Reviewer: This section reads much more clearly now, nicely done. But, for the paragraph after the listing of warnings, it should be "the *above* warnings" instead of "the *following* warnings".

```suggestion
  Normally, the above warnings are filtered by the default :ref:`warning
```

  filters <describing-warning-filters>`.

  It behaves as if the :option:`-W default <-W>` command line option is used.

  Use the :option:`-W error <-W>` command line option or set
  the :envvar:`PYTHONWARNINGS` environment variable to ``error`` to treat warnings
  as errors.

* Install debug hooks on memory allocators to check for:

  * Buffer underflow
  * Buffer overflow
  * Memory allocator API violation

>   * Usage of the GIL

Reviewer: Regarding "Usage of the GIL", should that specifically be "Unsafe usage of the GIL"? For example, do the debugs hooks trigger when:

1) Correctly using the `PyInterpreterState`, `PyThreadState`, and `PyGilState` APIs, exactly as the [C-API documentation describes](https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock)?

2) Unsafely/incorrectly using the GIL?

If it's (1), the current version is correct, but if it's (2), I'd recommend changing it to the following:

```suggestion
  * Unsafe usage of the GIL
```

The current version seems to imply that attempting to access or use the GIL in *any* form triggers the debug hooks. Is this true?

  the :const:`SIGSEGV`, :const:`SIGFPE`, :const:`SIGABRT`, :const:`SIGBUS` and
  :const:`SIGILL` signals to dump the Python traceback on a crash.

  It behaves as if the :option:`-X faulthandler <-X>` command line option is
  used or if the :envvar:`PYTHONFAULTHANDLER` environment variable is set to
  ``1``.

* Enable :ref:`asyncio debug mode <asyncio-debug-mode>`. For example,
  :mod:`asyncio` checks for coroutines that were not awaited and logs them.

  It behaves as if the :envvar:`PYTHONASYNCIODEBUG` environment variable is set
  to ``1``.

* Check the *encoding* and *errors* arguments for string encoding and decoding
  operations. Examples: :func:`open`, :meth:`str.encode` and
  :meth:`bytes.decode`.

  By default, for best performance, the *errors* argument is only checked at
  the first encoding/decoding error and the *encoding* argument is sometimes
  ignored for empty strings.

* The :class:`io.IOBase` destructor logs ``close()`` exceptions.
* Set the :attr:`~sys.flags.dev_mode` attribute of :attr:`sys.flags` to
  ``True``.

The Python Development Mode does not enable the :mod:`tracemalloc` module by default,

> because the overhead (performance and memory) would be too large. Enabling

Reviewer: ```suggestion
because the overhead cost (to performance and memory) would be too large. Enabling
```

the :mod:`tracemalloc` module provides additional information on the origin
of some errors. For example, :exc:`ResourceWarning` logs the traceback where
the resource was allocated, and a buffer overflow error logs the traceback
where the memory block was allocated.

> The Pyhon Development Mode does not prevent the :option:`-O` command line option to

Reviewer: ```suggestion
The Python Development Mode does not prevent the :option:`-O` command line option from
```

Also, see my previous comment on the next line: https://github.com/python/cpython/pull/18132/files#r370025635

    269

Enabling the Python Development Mode displays a :exc:`ResourceWarning` warning:

.. code-block:: shell-session

    $ python3 -X dev script.py README.txt
    269
    script.py:10: ResourceWarning: unclosed file <_io.TextIOWrapper name='README.rst' mode='r' encoding='UTF-8'>
      main()
    ResourceWarning: Enable tracemalloc to get the object allocation traceback

In addition, enabling :mod:`tracemalloc` shows the line where the file was
opened:

.. code-block:: shell-session

    $ python3 -X dev -X tracemalloc=5 script.py README.rst
    269
    script.py:10: ResourceWarning: unclosed file <_io.TextIOWrapper name='README.rst' mode='r' encoding='UTF-8'>
      main()
    Object allocated at (most recent call last):
      File "script.py", lineno 10
        main()
      File "script.py", lineno 4
        fp = open(sys.argv[1])

The fix is to close explicitly the file. Example using a context manager::

    def main():
        # Close the file explicitly when exiting the with block
        with open(sys.argv[1]) as fp:
            nlines = len(fp.readlines())
        print(nlines)

Not closing a resource explicitly can leave a resource open for way longer than

> expected; it can cause severe issues upon exiting the Python exit. It is bad in

Reviewer: ```suggestion
expected; it can cause severe issues upon exiting the Python interpreter. It is bad in
```

Author: I removed exit instead.
